### ISSUE TITLE
[BUG] Creating/updating cluster with RKE1 template containing private registry with empty password causes panic

### ISSUE THREAD

**Issue by bfbachmann** [rancher/rancher#41662]
**Rancher Server Setup**
- Rancher version: v2.7.2
- Installation option (Docker install/Helm Chart): Docker
- Proxy/Cert Details: Default for Rancher in Docker

**Information about the Cluster**
- Kubernetes version: Any (not relevant)
- Cluster Type (Local/Downstream): Downstream (but not relevant)
   - If downstream, what type of cluster? (Custom/Imported or specify provider for Hosted/Infrastructure Provider):

**User Information**
- What is the role of the user logged in? Admin

**Describe the bug**
Creating or updating a cluster with an RKE1 template with private registry enabled and an empty password will cause Rancher to panic. I think panic happens at [cluster_store.go:437](https://github.com/rancher/rancher/blob/ff2e09449efcfedfee865ccc67aa1b5dcab7183c/pkg/api/norman/store/cluster/cluster_store.go#L437) because of a forced type assertion. I think that when the value for the registry `password` is empty, it probably appears as `nil` in the `registries` map, and the panic happens because we're attempting to assert `nil` is of type `string`.

**To Reproduce**
1. In the Rancher UI, navigate to `Cluster Management`. Click the `RKE1 Configuration` dropdown on the left panel and then click `RKE Templates`. 
2. Create `Add template` in the top right and give it a name. Scroll down to the `Private Registry` section of the template and select `Enabled`. Put a dummy URL (can be anything) and User in the private registry config, but make sure to leave the password empty (aside: I don't know why anyone would set up a private registry with an empty password, but apparently they do). Make sure to also toggle the `Allow user override?` options next to `Private Registry`, `URL`, `User`, and `Password` so they can be overridden. 
3. Click the `Create` button at the bottom of the template.
4. Navigate back to `Clusters` and start creating a new RKE1 cluster (I'm using a DigitalOcean node-provisioned cluster, but I don't think that matters). Give it a name and a node with all 3 roles using any node template. Under `Cluster Options`, make sure you check the `Use an existing RKE Template and revision` option.
5. Under `Cluster Options`, select the template and revision you created in step 3. You should see that the `Private Registry` section of the cluster config now has your dummy registry URL, user, and empty password.
6. Click `Create` to create the cluster. You should see an empty red error box show up just above the `Cancel` and `Create` buttons. If you check the Rancher logs, you'll see a panic at [cluster_store.go:437](https://github.com/rancher/rancher/blob/ff2e09449efcfedfee865ccc67aa1b5dcab7183c/pkg/api/norman/store/cluster/cluster_store.go#L437).

**Result**
The cluster cannot be created.

**Expected Result**
The cluster is created.

**Screenshots**
<img width="1134" alt="image" src="https://github.com/rancher/rancher/assets/9647946/0c0b3bbe-4b2f-4a0a-b171-c71150708178">

**Additional context**
SURE-6366

Update on 2/8/2024 by @jiaqiluo:

In step 5, even if you enter a password you will still run into this bug. The root cause is the same.

**Comment by stefanlasiewski** [comment]
Hi Bruno,

I'm the one who reported this to support. I'll clarify a few bits:

> Result
> The cluster cannot be created.

Note that we are using this against live clusters, not a new cluster. In our experience, RKE will apply the template to the live cluster even if the Rancher UI shows an blank red error. That is the gist of the problem. If there was an error or if I made a mistake, then the Rancher UI should print an error message.

Under `To Reproduce`, step 2 you say:

> (aside: I don't know why anyone would set up a private registry with an empty password, but apparently they do).

From what I can tell, Rancher actually empties out that password field. An empty password is the default behavior.
I'll include some steps to reproduce with screenshots. I'm only documenting this for completeness and to show why the password was blank to begin with-- Fixing this bug would be out of scope of my main request. My main concern is that Rancher throws a blank error, and proceeds to change the cluster, as discussed above.

1. I navigate to ` ≡` > `Cluster Management` > `RKE1 Configuration`  > `RKE Templates` > select a template > `Clone Revision`
1. Scroll down to `Private Registry`
2. Notice that the password field is blank by default. 
3. Add a password. Your screen will look something like this screenshot.
![image]![image](https://github.com/rancher/rancher/assets/142752/93aaa260-d2e3-450e-b8a9-a8878998ec0a)
4. Click save.
5. View the saved template. Notice that the password field says `n/a`. That suggests there is no password.
6. Clone the template again and scroll down to `Private Registry`. The password field is still blank-- the password that I added 1 minute ago is not there. Cancel out of this dialog.
![image](https://github.com/rancher/rancher/assets/142752/b9c44a96-f966-4842-bce8-f164a456a80d)
7. Try to use the template for a cluster. Note that the password field is empty there also.

**Comment by snasovich** [comment]
@stefanlasiewski , thank you for additional clarifications and apologies for late response.
So it is a fair assessment that your main issue issue is more about password not saved in the template altogether than the fact that creating a cluster with an empty private registry fails?

**Comment by bfbachmann** [comment]
@stefanlasiewski Thanks for the detailed info. I believe the primary problem you're describing is this: https://github.com/rancher/rancher/issues/38819.

There is also a secondary issue where cloning a template with private registry without updating registry credentials results in empty registry credentials. I've filed an issue for this [here](https://github.com/rancher/rancher/issues/43361).

**Comment by stefanlasiewski** [comment]
Thanks @bfbachmann . Yes, I think I've also run into https://github.com/rancher/rancher/issues/43361 myself.

**Comment by Jono-SUSE-Rancher** [comment]
Created a backport issue for v2.8x-next1

**Comment by jiaqiluo** [comment]
can be validated in the latest v2.9-head

**Comment by markusewalker** [comment]
This issue will need an RC to properly test that this is addressed. In the meantime, QA will work on drafting a test plan + reproducing the original issue.

**Comment by markusewalker** [comment]
**QA TEST PLAN**

|#|Scenario|Test Case|
|---|---|---|
|1|Override RKE template that has an unauthenticated registry with an unauthenticated registry|Create an RKE template with an unauthenticated registry, provision a downstream RKE cluster that overrides the template with the same unauthenticated registry|
|2|Override RKE template that has an unauthenticated registry with an authenticated registry|Create an RKE template with an unauthenticated registry, provision a downstream RKE cluster that overrides the template with an authenticated registry|
|3|Override RKE template that has an authenticated registry with an unauthenticated registry|Create an RKE template with an authenticated registry, provision a downstream RKE cluster that override the template with an unauthenticated registry|
|4|Override RKE template that has an authenticated registry with an authenticated registry|Create an RKE template with an authenticated registry, provision a downstream RKE cluster that override the template with the same authenticated registry|

**Comment by markusewalker** [comment]
Reproduced scenario 1 of the test plan. See details below:

**ENVIRONMENT DETAILS**
- Rancher install: Docker
- Rancher version: `v2.8.2`

**TEST RESULT**

|#|Scenario|Test Case|Result
|---|---|---|---|
|1|Override RKE template that has an unauthenticated registry with an unauthenticated registry|Create an RKE template with an unauthenticated registry, provision a downstream RKE cluster that overrides the template with the same unauthenticated registry|:x:|

**REPRODUCTION STEPS**
1. Created an RKE template with a private registry that has an unauthenticated registry, i.e. no password.
2. Attempted to provision a downstream RKE cluster using that RKE template, but noted the following in the UI:
![image](https://github.com/rancher/rancher/assets/52591819/9f1d8571-be12-4174-910d-2185c2994caa)
3. Viewed the logs in my client machine that setup Rancher and noted the following panic:
```
2024/02/26 20:57:39 [ERROR] Panic serving api request: 
goroutine 40105 [running]:
runtime/debug.Stack()
	/usr/lib64/go/1.20/src/runtime/debug/stack.go:24 +0x65
github.com/rancher/norman/api.(*Server).ServeHTTP.func1()
	/go/pkg/mod/github.com/rancher/norman@v0.0.0-20240205154641-a6a6cf569608/api/server.go:175 +0x75
panic({0x421fde0, 0xc007103050})
	/usr/lib64/go/1.20/src/runtime/panic.go:884 +0x213
github.com/rancher/rancher/pkg/api/norman/store/cluster.loadDataFromTemplate(0xc0084ba500, 0x8153920?, 0x4bb2e8e?, 0xf?, 0x0, {0x5594960?, 0xc000efac78?})
	/go/src/github.com/rancher/rancher/pkg/api/norman/store/cluster/cluster_store.go:437 +0x218b
github.com/rancher/rancher/pkg/api/norman/store/cluster.(*Store).Create(0xc00075c180, 0xc001955320, 0xc00153afa8?, 0xc0089511a0)
	/go/src/github.com/rancher/rancher/pkg/api/norman/store/cluster/cluster_store.go:261 +0x23d
github.com/rancher/norman/store/wrapper.(*StoreWrapper).Create(0xc0004bd530, 0xc001955320, 0x314fd2d?, 0xc0089511a0)
	/go/pkg/mod/github.com/rancher/norman@v0.0.0-20240205154641-a6a6cf569608/store/wrapper/wrapper.go:69 +0x162
github.com/rancher/norman/api/handler.CreateHandler(0xc001955320, 0xc00e3ee590?)
	/go/pkg/mod/github.com/rancher/norman@v0.0.0-20240205154641-a6a6cf569608/api/handler/create.go:23 +0x58
github.com/rancher/norman/api.(*Server).handle(0xc0008f6210, {0x55b7280?, 0xc00e3ee590?}, 0xc00f42c968?)
	/go/pkg/mod/github.com/rancher/norman@v0.0.0-20240205154641-a6a6cf569608/api/server.go:249 +0x334
github.com/rancher/norman/api.(*Server).ServeHTTP(0x46a13a0?, {0x55b7280?, 0xc00e3ee590?}, 0x7f8c61f632c0?)
	/go/pkg/mod/github.com/rancher/norman@v0.0.0-20240205154641-a6a6cf569608/api/server.go:180 +0x71
github.com/rancher/apiserver/pkg/middleware.ContentType.func1({0x55a2370?, 0xc00c3489e0?}, 0x1?)
	/go/pkg/mod/github.com/rancher/apiserver@v0.0.0-20240205154815-a3b9e3721c1b/pkg/middleware/content.go:33 +0x7b
net/http.HandlerFunc.ServeHTTP(0x5582be0?, {0x55a2370?, 0xc00c3489e0?}, 0x4?)
	/usr/lib64/go/1.20/src/net/http/server.go:2122 +0x2f
github.com/rancher/apiserver/pkg/middleware.Gzip.func1({0x55b6890?, 0xc0004fa1c0}, 0xc00fd92b00)
	/go/pkg/mod/github.com/rancher/apiserver@v0.0.0-20240205154815-a3b9e3721c1b/pkg/middleware/gzip.go:54 +0x19f
net/http.HandlerFunc.ServeHTTP(0xc00f42cbe0?, {0x55b6890?, 0xc0004fa1c0?}, 0x1a?)
	/usr/lib64/go/1.20/src/net/http/server.go:2122 +0x2f
github.com/rancher/rancher/pkg/auth/requests.authHeaderHandler.ServeHTTP({{0x5586ca0?, 0xc003223848?}}, {0x55b6890, 0xc0004fa1c0}, 0xc00fd92b00)
	/go/src/github.com/rancher/rancher/pkg/auth/requests/filter.go:64 +0x715
github.com/rancher/rancher/pkg/rbac.NewAccessControlHandler.func1.1({0x55b6890, 0xc0004fa1c0}, 0xc00fd92a00)
	/go/src/github.com/rancher/rancher/pkg/rbac/user_based.go:33 +0x18d
net/http.HandlerFunc.ServeHTTP(0x55b8fb0?, {0x55b6890?, 0xc0004fa1c0?}, 0x5563400?)
	/usr/lib64/go/1.20/src/net/http/server.go:2122 +0x2f
github.com/rancher/steve/pkg/auth.ToMiddleware.func1.1({0x55b6890, 0xc0004fa1c0}, 0xc00fd92900)
	/go/pkg/mod/github.com/rancher/steve@v0.0.0-20231016202603-993540401906/pkg/auth/filter.go:172 +0x30d
net/http.HandlerFunc.ServeHTTP(0xc00fd92600?, {0x55b6890?, 0xc0004fa1c0?}, 0x479a8a0?)
	/usr/lib64/go/1.20/src/net/http/server.go:2122 +0x2f
github.com/gorilla/mux.(*Router).ServeHTTP(0xc00060d680, {0x55b6890, 0xc0004fa1c0}, 0xc00fd92500)
	/go/pkg/mod/github.com/gorilla/mux@v1.8.0/mux.go:210 +0x1cf
github.com/gorilla/mux.(*Router).ServeHTTP(0xc00060d500, {0x55b6890, 0xc0004fa1c0}, 0xc00fd92300)
	/go/pkg/mod/github.com/gorilla/mux@v1.8.0/mux.go:210 +0x1cf
github.com/rancher/rancher/pkg/multiclustermanager.(*DeferredServer).Middleware.func1({0x55b6890, 0xc0004fa1c0}, 0xc0089506f0?)
	/go/src/github.com/rancher/rancher/pkg/multiclustermanager/deferred.go:113 +0x78
net/http.HandlerFunc.ServeHTTP(0xc00fd92200?, {0x55b6890?, 0xc0004fa1c0?}, 0x479a8a0?)
	/usr/lib64/go/1.20/src/net/http/server.go:2122 +0x2f
github.com/gorilla/mux.(*Router).ServeHTTP(0xc00100d080, {0x55b6890, 0xc0004fa1c0}, 0xc00fd92100)
	/go/pkg/mod/github.com/gorilla/mux@v1.8.0/mux.go:210 +0x1cf
github.com/gorilla/mux.(*Router).ServeHTTP(0xc00b784000, {0x55b6890, 0xc0004fa1c0}, 0xc0051f3f00)
	/go/pkg/mod/github.com/gorilla/mux@v1.8.0/mux.go:210 +0x1cf
github.com/rancher/rancher/pkg/api/steve/aggregation.(*aggregationHandler).Middleware.func1({0x55b6890, 0xc0004fa1c0}, 0xc008950540?)
	/go/src/github.com/rancher/rancher/pkg/api/steve/aggregation/aggregation.go:52 +0x55
net/http.HandlerFunc.ServeHTTP(0xc0051f3e00?, {0x55b6890?, 0xc0004fa1c0?}, 0xc00f42d368?)
	/usr/lib64/go/1.20/src/net/http/server.go:2122 +0x2f
github.com/gorilla/mux.(*Router).ServeHTTP(0xc00100cf00, {0x55b6890, 0xc0004fa1c0}, 0xc0051f3d00)
	/go/pkg/mod/github.com/gorilla/mux@v1.8.0/mux.go:210 +0x1cf
github.com/rancher/rancher/pkg/api/steve/proxy.NewProxyMiddleware.func1.1({0x55b6890?, 0xc0004fa1c0?}, 0xc00503c970?)
	/go/src/github.com/rancher/rancher/pkg/api/steve/proxy/proxy.go:80 +0x68
net/http.HandlerFunc.ServeHTTP(0x39d43e9?, {0x55b6890?, 0xc0004fa1c0?}, 0xc00f8eb2f4?)
	/usr/lib64/go/1.20/src/net/http/server.go:2122 +0x2f
github.com/rancher/rancher/pkg/api/steve/proxy.RewriteLocalCluster.func1({0x55b6890, 0xc0004fa1c0}, 0xc0051f3d00)
	/go/src/github.com/rancher/rancher/pkg/api/steve/proxy/proxy.go:46 +0x177
net/http.HandlerFunc.ServeHTTP(0xc00e3ee520?, {0x55b6890?, 0xc0004fa1c0?}, 0x4b83719?)
	/usr/lib64/go/1.20/src/net/http/server.go:2122 +0x2f
github.com/rancher/rancher/pkg/websocket.websocketHandler.ServeHTTP({{0x5586ca0?, 0xc0006e7188?}}, {0x55b6890, 0xc0004fa1c0}, 0xc0051f3d00)
	/go/src/github.com/rancher/rancher/pkg/websocket/handler.go:35 +0xa8
github.com/rancher/apiserver/pkg/middleware.NoCache.func1({0x55b6890, 0xc0004fa1c0}, 0xc010f60420?)
	/go/pkg/mod/github.com/rancher/apiserver@v0.0.0-20240205154815-a3b9e3721c1b/pkg/middleware/cache.go:33 +0xf4
net/http.HandlerFunc.ServeHTTP(0x46b36a0?, {0x55b6890?, 0xc0004fa1c0?}, 0x16?)
	/usr/lib64/go/1.20/src/net/http/server.go:2122 +0x2f
github.com/rancher/apiserver/pkg/middleware.ContentTypeOptions.func1({0x55b6890, 0xc0004fa1c0}, 0xc010f60420?)
	/go/pkg/mod/github.com/rancher/apiserver@v0.0.0-20240205154815-a3b9e3721c1b/pkg/middleware/cache.go:47 +0xf4
net/http.HandlerFunc.ServeHTTP(0x46b36a0?, {0x55b6890?, 0xc0004fa1c0?}, 0x11?)
	/usr/lib64/go/1.20/src/net/http/server.go:2122 +0x2f
github.com/rancher/rancher/pkg/auth.SetXAPICattleAuthHeader.func1({0x55b6890, 0xc0004fa1c0}, 0xc0051f3d00)
	/go/src/github.com/rancher/rancher/pkg/auth/server.go:180 +0x19c
net/http.HandlerFunc.ServeHTTP(0x81b8ec0?, {0x55b6890?, 0xc0004fa1c0?}, 0xc00f42d888?)
	/usr/lib64/go/1.20/src/net/http/server.go:2122 +0x2f
github.com/rancher/rancher/pkg/auth/audit.auditHandler.ServeHTTP({{0x5586ca0, 0xc0006e71d0}, 0x0, 0xc001c70aa0, 0xc008a18000, 0xc008130280}, {0x55b6890?, 0xc0004fa1c0?}, 0xc0051f3d00?)
	/go/src/github.com/rancher/rancher/pkg/auth/audit/filter.go:67 +0x566
github.com/rancher/steve/pkg/auth.ToMiddleware.func1.1({0x55b6890, 0xc0004fa1c0}, 0xc0051f3b00)
	/go/pkg/mod/github.com/rancher/steve@v0.0.0-20231016202603-993540401906/pkg/auth/filter.go:172 +0x30d
net/http.HandlerFunc.ServeHTTP(0xe0?, {0x55b6890?, 0xc0004fa1c0?}, 0xc00f42db70?)
	/usr/lib64/go/1.20/src/net/http/server.go:2122 +0x2f
github.com/rancher/dynamiclistener/server.wrapHandler.func1({0x55b6890, 0xc0004fa1c0}, 0xc00f42dae0?)
	/go/pkg/mod/github.com/rancher/dynamiclistener@v0.3.6/server/server.go:199 +0x6e
net/http.HandlerFunc.ServeHTTP(0x55e609?, {0x55b6890?, 0xc0004fa1c0?}, 0x3b?)
	/usr/lib64/go/1.20/src/net/http/server.go:2122 +0x2f
github.com/rancher/dynamiclistener.HTTPRedirect.func1({0x55b6890, 0xc0004fa1c0}, 0xc0051f3b00)
	/go/pkg/mod/github.com/rancher/dynamiclistener@v0.3.6/redirect.go:20 +0x1e4
net/http.HandlerFunc.ServeHTTP(0xc00e339151?, {0x55b6890?, 0xc0004fa1c0?}, 0x72a5d4?)
	/usr/lib64/go/1.20/src/net/http/server.go:2122 +0x2f
net/http.serverHandler.ServeHTTP({0x559c160?}, {0x55b6890, 0xc0004fa1c0}, 0xc0051f3b00)
	/usr/lib64/go/1.20/src/net/http/server.go:2936 +0x316
net/http.(*conn).serve(0xc00ecc0630, {0x55b8fb0, 0xc0091ae1b0})
	/usr/lib64/go/1.20/src/net/http/server.go:1995 +0x612
created by net/http.(*Server).Serve
	/usr/lib64/go/1.20/src/net/http/server.go:3089 +0x5ed
```

**Comment by markusewalker** [comment]
This issue is waiting for an RC to test. Once given, QA can proceed with validating this issue.

**Comment by markusewalker** [comment]
Validated that this is addressed on `v2.9.0-alpha1`. See details below:

**ENVIRONMENT DETAILS**
- Rancher install: Docker
- Rancher version: `v2.9.0-alpha1`

**TEST RESULT**
|#|Scenario|Test Case|Result
|---|---|---|---|
|1|Override RKE template that has an unauthenticated registry with an unauthenticated registry|Create an RKE template with an unauthenticated registry, provision a downstream RKE cluster that overrides the template with the same unauthenticated registry|:white_check_mark:|
|2|Override RKE template that has an authenticated registry with an unauthenticated registry|Create an RKE template with an authenticated registry, provision a downstream RKE cluster that override the template with an unauthenticated registry|:white_check_mark:|
|3|Override RKE template that has an authenticated registry with an authenticated registry|Create an RKE template with an authenticated registry, provision a downstream RKE cluster that override the template with the same authenticated registry|:white_check_mark:|

**VALIDATION STEPS**

**Scenario 1**
1. Created an RKE template with a private registry that has an unauthenticated registry.
2. Provisioned a downstream RKE cluster using that RKE template.
     -  Override the registry with the same registry.
3. Validated that the cluster and nodes came up `Active`.

___________

**Scenario 2**
1. Created an RKE template with a private registry that has an authenticated registry.
2. Provisioned a downstream RKE cluster using that RKE template.
     - Override the registry with the same registry
3. Validated that the cluster and nodes came up `Active`.

_________

**Scenario 3**
1. Created an RKE template with a private registry that has an authenticated registry.
2. Provisioned a downstream RKE cluster using that RKE template.
     -Override the registry to use an unauthenticated registry.
3. Validated that the cluster and nodes came up `Active`.
